plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,238 @@
+
+    str_len = 62 + flb_sds_len(ctx->db_table);
+
+    query = flb_malloc(str_len);
+    snprintf(query, str_len, "CREATE TABLE IF NOT EXISTS %s (data json);", ctx->db_table);
+    res = PQexec(ctx->conn, query);
+    flb_free(query);
+
+    if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+        flb_error("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+        PQclear(res);
+        return -1;
+    }
+
+    PQclear(res);
+
+    flb_info("[out_pgsql] switching postgresql connection to non-blocking mode");
+    if(PQsetnonblocking(ctx->conn, 1) != 0) {
+        flb_error("[out_pgsql] non-blocking mode not set, not good");
+        return -1;
+    }
+
+    return 0;
+}
+
+static void cb_pgsql_flush(const void *data, size_t bytes,
+                            const char *tag, int tag_len,
+                            struct flb_input_instance *i_ins,
+                            void *out_context,
+                            struct flb_config *config)
+{
+    struct flb_pgsql_config *ctx = out_context;
+    flb_sds_t json;
+    PGresult *res = NULL;
+    int result;
+    char *query = NULL;
+    size_t str_len;
+
+    if(PQconsumeInput(ctx->conn) == 0 && PQisBusy(ctx->conn) == 1) {
+        flb_info("[out_pgsql] Some command may still be running");
+    }
+
+    /*
+      PQreset()
+      This function will close the connection to the server and attempt to reestablish a new connection
+      to the same server, using all the same parameters previously used. This might be useful for error
+      recovery if a working connection is lost.
+     */
+    if(PQstatus(ctx->conn) != CONNECTION_OK) {
+        PQreset(ctx->conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+
+    json = flb_pack_msgpack_to_json_format(data, bytes,
+                                           FLB_PACK_JSON_FORMAT_JSON,
+                                           FLB_PACK_JSON_DATE_DOUBLE,
+                                           ctx->timestamp_key);
+
+    str_len = 115 + flb_sds_len(json);
+
+    query = flb_malloc(str_len);
+    snprintf(query, str_len,
+             "with json_pack as (select json_array_elements('%s') as json_pack )INSERT INTO %s select * from json_pack;",
+             json, ctx->db_table);
+
+    result = PQsendQuery(ctx->conn, query);
+    flb_free(query);
+    flb_sds_destroy(json);
+
+    PQflush(ctx->conn);
+
+    if(PQisBusy(ctx->conn) == 0) {
+        res = PQgetResult(ctx->conn);
+        if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+            flb_warn("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+        }
+        PQclear(res);
+    }
+
+    FLB_OUTPUT_RETURN(FLB_OK);
+}
+
+static int cb_pgsql_exit(void *data, struct flb_config *config)
+{
+    struct flb_pgsql_config *ctx = data;
+
+    if(!ctx){
+        return 0;
+    }
+
+    pgsql_conf_destroy(ctx);
+
+    return 0;
+}
+
+struct flb_output_plugin out_pgsql_plugin = {
+    .name         = "pgsql",
+    .description  = "PostgreSQL database",
```

Comment:
Suggestion: "Output to PostgreSQL"

Reply:
I thought the same but all the other plugins don't use the "output" word explicitly, do you think that we should change this? it looks redundant to me since we're talking about out_pgsql so the message it's more like to clarify the "pgsql" word rather than explicitly say it's about an output plugin, but yeah maybe we need to chat about this

Reply:
hmmm

```
  es                    Elasticsearch
  forward               Forward (Fluentd protocol)
  http                  HTTP Output
  influxdb              InfluxDB Time Series
  kafka-rest            Kafka REST Proxy
  nats                  NATS Server
```
I think ```PostgreSQL``` should be enough...

Reply:
Yeap but the'res a good reason to put Output in the prefix... what if we create an input plugin later? =)

Reply:
having the same name is fine, take a look at the case of ```forward``` plugin:

```
$ fluent-bit -h

Inputs
  ...
  forward               Fluentd in-forward

Outputs
  ...
  forward               Forward (Fluentd protocol)

Filters
  ...

```

Reply:
note that input plugin at a file system level are prefixed with ```in_``` while outputs are ```out_```

Reply:
ho! ok! make sense to me now